converter/stitching/stitcher.go

```diff
@@ -124,41 +126,55 @@ func (s *stitcher) stitchFlow(flow input.Flow) error {
 		if s.shouldMerge(&newSessAgg, &oldSessAgg) {
 			matchFound = true
 
-			//merge the session aggregates (adding data counts etc)
-			err = newSessAgg.Merge(&oldSessAgg)
-			if err != nil {
-				return errors.Wrapf(err, "cannot merge session\n%+v\nwith\n%+v", &newSessAgg, &oldSessAgg)
+			f_FlowStartMilliseconds := newSessAgg.FlowStartMilliseconds()
+			nextMatch_FlowStartMilliseconds := newSessAgg.FlowStartMilliseconds()
+			f_FlowEndMilliseconds := newSessAgg.FlowEndMilliseconds()
+			nextMatch_FlowEndMilliseconds := oldSessAgg.FlowEndMilliseconds()
+
+			var diff1 = f_FlowEndMilliseconds - nextMatch_FlowEndMilliseconds
+			if diff1 < 0 {
+				diff1 *= -1
 			}
-			//if both sides of the session have been filled, write it out
-			if newSessAgg.FilledFromSourceA && newSessAgg.FilledFromSourceB {
-				err := s.matcher.Remove(&oldSessAgg)
-				if err != nil {
-					return errors.Wrap(err, "could not remove old session aggregate")
-				}
-				s.sessionsOut <- &newSessAgg
-			} else {
-				//otherwise update the database
-				newSessAgg.MatcherID = oldSessAgg.MatcherID
-				err := s.matcher.Update(&newSessAgg)
-				if err != nil {
-					return errors.Wrap(err, "could not update existing session aggregate")
-				}
+			var diff2 = f_FlowStartMilliseconds - nextMatch_FlowStartMilliseconds
+			if diff2 < 0 {
+				diff2 *= -1
 			}
-		}
-	}
 
-	//if there's an error other than not found, return it up
-	if oldSessAggIter.Err() != nil /*&& oldSessAggIter.Err() != mgo.ErrNotFound*/ {
-		return errors.Wrap(oldSessAggIter.Err(), "could not find all matching session aggregates")
+			newMatchCost := diff1 + diff2
+
+			if newMatchCost < matchCost {
+				matchCost = newMatchCost
+				matchAgg = oldSessAgg
+			}
+		}
 	}
 
-	//no matching unstitched session found
-	if !matchFound {
+	if matchFound {
+		err = newSessAgg.Merge(&oldSessAgg)
```

Comment:
Sorry, missed this. every instance of oldSessAgg in the if matchFound block should be matchAgg. oldSessAgg will be set to the last matching record we checked, we want it to be the optimal aggregate instead.